Perform file validation with file extension first, then file mime type

kandula-aswin opened the pull request:
resolves #1584

marksreeves commented on wcomponents-core/src/main/java/com/github/bordertech/wcomponents/WFileWidget.java:
```diff
@@ -77,9 +79,19 @@ public void setFileTypes(final String[] types) {
 	/**
-	 * Set each file type as a valid file mime type to be accepted by the WMultiFileWidget.
+	 * Determines the file types accepted by this widget. Note that duplicates are not allowed and these are not case
+	 * sensitive. A file type may be one of the following:
```
These *are* the only MIME types. The fourth item in the list covers that. The list of acceptable content mirrors the HTML spec for the `accepts` attribute.

kandula-aswin commented:
@jonathanaustin 
* updated javadoc to reflect the new implementation of validating the file types (both MIME and Ext)
* introduced new method getMimeType(), allowing developers to implement their own validation
Please review

marksreeves commented on wcomponents-core/src/main/java/com/github/bordertech/wcomponents/WFileWidget.java:
```diff
@@ -80,16 +86,27 @@ public void setFileTypes(final String[] types) {
-	 * extension are accepted).</li>
+	 * <li>MIME type </li>
+	 * <li>Extension</li>
 	 * </ul>
```
Restore the original text of this doclet as it is **more correct** than the text with which you have replaced it. @jonathanaustin was incorrect in his earlier assertion.

kandula-aswin replied:
Perhaps we 3 should meet and reach a concences on the solution, I know I had seperate conversations and the opinion differed slightly.

marksreeves commented on wcomponents-core/src/main/java/com/github/bordertech/wcomponents/WFileWidget.java:
```diff
@@ -80,16 +86,27 @@ public void setFileTypes(final String[] types) {
+	 * <ul>
+	 * <li><i>.txt</i> - indicates any files with extension <b>txt</b> are accepted.</li>
+	 * <li><i>.jpg</i> - indicates any files with extension <b>jpg</b> are accepted.</li>
+	 * </ul>
```
As part of the restoration of the original text you may want to keep this as further elucidation but it is generally redundant. If you do keep this content then remove all `u` elements as they add no value and reduce usability; replace all `i` elements with `em` elements; and replace all `b` elements with `strong` elements.

kandula-aswin replied:
ok

marksreeves commented on wcomponents-core/src/main/java/com/github/bordertech/wcomponents/WFileWidget.java:
```diff
@@ -358,6 +375,20 @@ public FileItemWrap getValue() {
+		}
+		return null;
+	}
+
```
This function **always** returns `null`. Is that correct?

kandula-aswin replied:
No, if the widget has a file, then it's mime will be returned, otherwise null. 
I had added [additional class level javadoc](https://github.com/BorderTech/wcomponents/blob/4e6e825546737773489ce40ae9f377734e0e50c4/wcomponents-core/src/main/java/com/github/bordertech/wcomponents/WFileWidget.java#L38) indicating the new API, allowing the client to perform custom mime validations

marksreeves commented on wcomponents-core/src/main/java/com/github/bordertech/wcomponents/WMultiFileWidget.java:
```diff
@@ -159,6 +165,21 @@ public FileWidgetUpload getFile(final String fileId) {
+			FileUtil.getFileMimeType(getFile(fileId).getFile());
+		}
+		return null;
+	}
```
This function **always** returns `null`. It appears to be an error.

kandula-aswin replied:
No, if the widget has a file, then it's mime will be returned, otherwise null.
I had added [additional class level javadoc](https://github.com/BorderTech/wcomponents/blob/4e6e825546737773489ce40ae9f377734e0e50c4/wcomponents-core/src/main/java/com/github/bordertech/wcomponents/WMultiFileWidget.java#L36) indicating the new API, allowing the client to perform custom mime validations

marksreeves replied:
Where are the unit tests?

marksreeves commented on wcomponents-core/src/main/java/com/github/bordertech/wcomponents/WMultiFileWidget.java:
```diff
@@ -208,16 +229,27 @@ public void setFileTypes(final String[] types) {
+	 * sensitive.<br>
+	 * The file type(s) can be either:
 	 * <ul>
-	 * <li>The string audio/* (Indicates that sound files are accepted.)</li>
```
Restore the original text of this list as it is **more correct** than that which replaced it.

marksreeves commented on wcomponents-core/src/main/java/com/github/bordertech/wcomponents/WMultiFileWidget.java:
```diff
@@ -208,16 +229,27 @@ public void setFileTypes(final String[] types) {
+	 * <ul>
+	 * <li><i>.txt</i> - indicates any files with extension <b>txt</b> are accepted.</li>
+	 * <li><i>.jpg</i> - indicates any files with extension <b>jpg</b> are accepted.</li>
+	 * </ul>
```
If keeping this extended elucidation then fix the markup as per the comment above.

kandula-aswin replied:
ok

marksreeves commented on wcomponents-core/src/main/java/com/github/bordertech/wcomponents/util/FileUtil.java:
```diff
@@ -68,18 +68,17 @@ public static boolean validateFileType(final FileItemWrap newFile, final List<St
 			}
 		}
 		// If extension match is unsucessful, then move to fileMimes list
 		boolean mimeMatch = false;
```
is the initial value strictly necessary?

kandula-aswin replied:
while strictly not necessary, if not initialised to `false`, I need to set to false 
* in `if` statement in `try` block,  
* `catch` block, 
* and add new else block after [line 93](https://github.com/BorderTech/wcomponents/blob/4e6e825546737773489ce40ae9f377734e0e50c4/wcomponents-core/src/main/java/com/github/bordertech/wcomponents/util/FileUtil.java#L93)

marksreeves replied:
ok

marksreeves commented on wcomponents-core/src/main/java/com/github/bordertech/wcomponents/util/FileUtil.java:
```diff
@@ -96,6 +95,21 @@ public static boolean validateFileType(final FileItemWrap newFile, final List<St
+	 * @throws IOException if the mime type detection fails.
+	 */
+	public static String getFileMimeType(final File file) throws IOException {
+		if (file != null) {
```
This is OK but given this is a public method don't you think it would be safer to throw an illegal argument exception if we don't have a file? After all, the MIME type of `null` is not `null`.

kandula-aswin replied:
Because it is dealing with File, I felt leaving `IOException` is appropriate. Both illegal argument exception and `IOException` are runtime exceptions, so the client need not catch.
But I accept to change to illegal argument exception if it makes more sense.

marksreeves commented on wcomponents-core/src/main/java/com/github/bordertech/wcomponents/util/FileUtil.java:
```diff
@@ -74,21 +74,17 @@ public static boolean validateFileType(final FileItemWrap newFile, final List<St
 		// If extension match is unsucessful, then move to fileMimes list
 		boolean mimeMatch = false;
 		if (!extMatch && fileMimes.size() > 0) {
-			try {
```
You could reduce the cyclometric complexity significantly by returning true if `extMatch` is `true` before you even get in here. as the final line returns `extMatch || mimeMatch`.

marksreeves commented on wcomponents-core/src/main/java/com/github/bordertech/wcomponents/util/FileUtil.java:
```diff
@@ -74,21 +74,17 @@ public static boolean validateFileType(final FileItemWrap newFile, final List<St
+			LOG.debug("File mime-type is: " + mimeType);
+			for (String fileMime : fileMimes) {
+				if (StringUtils.equals(mimeType, fileMime)) {
+					mimeMatch = true;
```
the function returns `true` if `mimeMatch` == `true` so you can significantly reduce cyclometric complexity by returning `true` rather than setting the member. This goes for all other of same type in this nested if group.

jonathanaustin commented on wcomponents-core/src/main/java/com/github/bordertech/wcomponents/util/FileUtil.java:
```diff
@@ -23,68 +24,108 @@
+				final Tika tika = new Tika();
+				return tika.detect(file.getInputStream());
+			} catch (IOException ex) {
+				LOG.error("Invalid file.");
```
Might be good to log more details.

kandula-aswin commented:
Merging PR after discussing with @marksreeves
